dcap/ql/src/quote_generator.rs

```diff
@@ -6,18 +6,51 @@
 //! functionality will return errors.
 
 use mc_sgx_core_sys_types::sgx_target_info_t;
-use mc_sgx_core_types::TargetInfo;
+use mc_sgx_core_types::{Quote, Report, ReportData, TargetInfo};
 use mc_sgx_dcap_types::Quote3Error;
 use mc_sgx_util::ResultInto;
 
-/// Provides behavior to generate quotes for enclaves
+/// Functionality to create quotes.  Implementers should only need to implement
+/// `report` as that will need to come from the application enclave.
 pub trait QuoteGenerator {
+    /// Report from an application enclave
+    ///
+    /// # Arguments
+    /// * `target_info` - The information for the target enclave which will
+    ///   cryptographically verify the report. This is usually from the
+    ///   QE(Quoting Enclave).
+    /// * `report_data` - Report data used to communicate between enclaves
+    fn report(
+        target_info: &TargetInfo,
+        report_data: Option<&ReportData>,
+    ) -> Result<Report, Quote3Error>;
+
     /// The target info of the QE(Quoting Enclave)
     fn target_info() -> Result<TargetInfo, Quote3Error> {
         let mut info = sgx_target_info_t::default();
         unsafe { mc_sgx_dcap_ql_sys::sgx_qe_get_target_info(&mut info) }.into_result()?;
         Ok(info.into())
     }
+
+    /// A new quote for the instance
+    ///
+    /// # Arguments
+    /// * `report_data` - Report data used to communicate between enclaves
+    fn quote(report_data: Option<&ReportData>) -> Result<Quote<Vec<u8>>, Quote3Error> {
+        let mut size = 0;
+        unsafe { mc_sgx_dcap_ql_sys::sgx_qe_get_quote_size(&mut size) }.into_result()?;
```

Comment:
I'm up for ideas on testing, I can continue to build the HW tests.  I don't like that we currently only have a manual way to run them right now.
I could also look at out of process quoting and maybe mocking those interfaces out for tests

Reply:
code changed so comment out of context.